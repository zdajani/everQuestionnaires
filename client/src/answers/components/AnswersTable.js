import React from 'react';
import PropTypes from 'prop-types';

const AnswersTable = ({questions, usersAnswers}) => {
  //returns questions
  const renderTableHeaders = () => {
    return questions.map(question => {
      return (
        <th key={question.id}>{question.name}</th>
      )
    })
  }

  const renderRows = () => {
    return usersAnswers.map((userAnswers, index) => {
      return (
        <tr key={index}>
          <th scope="row">{index + 1}</th>
          { renderRow(userAnswers.username, userAnswers.id )}
          { renderAnswers(userAnswers.answers) }
          { renderRow(new Date(userAnswers.date).toDateString()) }
        </tr>
      )
    })
  }

  const renderAnswers = (answers) => {
    return questions.map((question, index) => {
      const answer = answers.find(x => x.question_id === question.id);
      const response = answer ? answer.response : null

      return (
        renderRow(response, index)
      );
    });
  }

  const renderRow =(data, id) => {
    return (
        <td key={id}>{ data }</td>
    )
  }

  return (
    <div>
      <table className="table table-striped table-bordered table-responsive">
        <thead className="thead-inverse">
          <tr>
            <th>#</th>
            <th>Username</th>
            { renderTableHeaders() }
            <th>Date</th>
          </tr>
        </thead>
        <tbody>
          {renderRows()}
        </tbody>
      </table>
    </div>
  )
}

AnswersTable.propTypes = {
  questions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired
  }).isRequired),
  usersAnswers: PropTypes.arrayOf(PropTypes.shape({
    username: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    answers: PropTypes.arrayOf(PropTypes.shape({
      id:PropTypes.number.isRequired,
      response:PropTypes.string.isRequired,
      question_id: PropTypes.number.isRequired
    }).isRequired).isRequired
  }).isRequired)
};

export default AnswersTable;