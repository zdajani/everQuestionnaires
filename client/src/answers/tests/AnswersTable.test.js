import React from 'react';
import { shallow } from 'enzyme';
import AnswersTable from '../components/AnswersTable';
import { answersTableData } from './testData';
import toJson from 'enzyme-to-json';

const setup = () => {
  const component = shallow(
    <AnswersTable 
      questions={answersTableData.questions} 
      usersAnswers={answersTableData.usersAnswers} />
  );
    
  return {
    component
  };
}

describe('AnswersTable component', () => {
  const { component } = setup();
  
  it('renders self with questions and answers props', () => {
    expect(component.instance().props).toEqual(
      {
        questions: answersTableData.questions, 
        usersAnswers:answersTableData.usersAnswers
      })
      expect(toJson(component)).toMatchSnapshot();
  });
})