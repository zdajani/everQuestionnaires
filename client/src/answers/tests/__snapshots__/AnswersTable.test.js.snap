// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswersTable component renders self with questions and answers props 1`] = `
<div>
  <table
    className="table table-striped table-bordered table-responsive"
  >
    <thead
      className="thead-inverse"
    >
      <tr>
        <th>
          #
        </th>
        <th>
          Username
        </th>
        <th>
          Ragdoll traits
        </th>
        <th>
          Persian traits
        </th>
        <th>
          Date
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1
        </th>
        <td>
          user1
        </td>
        <td>
          Pellentesque in ipsum id orci porta dapibus.
        </td>
        <td>
          Vivamus suscipit tortor eget felis porttitor.
        </td>
        <td>
          Thu Aug 17 2017
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
